src/render/decorations/IHyprWindowDecoration.hpp

```diff
@@ -39,6 +39,12 @@ class IHyprWindowDecoration {
 
     virtual bool                     allowsInput();
 
+    virtual void                     dragWindowToDecoration(CWindow*, const Vector2D&);
+
+    virtual void                     clickDecoration(const Vector2D&);
+
+    virtual void                     dragFromDecoration(const Vector2D&);
```

Comment:
begin? end?

Reply:
what about?
```
virtual void                     onEndDragWindow(CWindow*, const Vector2D&);
virtual void                     onClickDecoration(const Vector2D&);
virtual void                     onBeginDragWindow(const Vector2D&);

```

Reply:
hm. A bit non-descriptive. 

What is `onBeginDragWindow`? on begin drag of the parent? the deco?

if we want to handle drags on the decoration, why not just `onMouseDown` `onMouseUp`?

The names should generally make it clear what the function is about.

Reply:
`onEndDragWindow` should be called when a window gets dragged onto a decorations (`movewindow` mouse dispatcher) (not what currently happens but TODO should get fixed before that is done)

`onBeginDragWindow` is called when `movewindow` is used but mouse position is inside a window decoration region

not really `onMouseDown` and `onMouseUp`

Reply:
We can make both into `onBeginDragOnDeco` `onEndDragOnDeco` with an argument like an enum `eDragDecoType` with stuff like `DRAG_TYPE_WINDOW` or stuff, and an `std::any` for data. In the enum, we can use comments to doc what data is passed via the any.

Reply:
> We can make both into `onBeginDragOnDeco` `onEndDragOnDeco` with an argument like an enum `eDragDecoType` with stuff like `DRAG_TYPE_WINDOW` or stuff, and an `std::any` for data. In the enum, we can use comments to doc what data is passed via the any.
> 
> Though we can also add a `DRAG_DATA_NONE`, and then document it could be a click as well, but at that point it's a `onDecoClickBegin` `onDecoClickEnd`
> 
> Some decos might want to respond to clicks in general, and currently they have to hook into events for that.

not sure what you mean, are you suggesting turning the 3 functions into 1 that handles all the events?

Reply:
no, more like
```
void onDecoClickBegin(const Vector2D& pos, const eClickDataType type, std::any data);
```

where `eClickDataType` could be `CLICK_DATA_NONE`, `CLICK_DATA_WINDOW`, etc. First would imply no data, second would imply `std::any data` has a `CWindow*`

Reply:
still don't get why/how would a click pass a `Cwindow*`, would this be used instead of the current `onBeginDragWindow`?

Reply:
generally my idea is to pass clicks to the decoration, so that in the future, for example, you can drag tabs around.

Unifying this with dropping a window would be nice.

Reply:
> generally my idea is to pass clicks to the decoration, so that in the future, for example, you can drag tabs around.
> 
> Unifying this with dropping a window would be nice.

wouldn't this be done by adding another drag mode in input manager?
that wouldn't need any extra arguments, except the mouse event type

Reply:
well no, these are more like "attachments" while drag modes are like resizes

Reply:
I'm pretty sure drag mode would have to be involved for dragging tabs from one group to another.

same group dragging should be doable by adding an mouse event param to `onClickDecoration`, and since it would be fully handled by the deco there would be no need to pass window as a parameter

Reply:
dont understand your point here.

Reply:
if you're clicking down on a decoration and releasing on mouse up or when leaving deco, that can fully done just by sending the mouse up/down and the mouse move events, no need for extra args

my point is, I still don't get what 
> ```
> void onDecoClickBegin(const Vector2D& pos, const eClickDataType type, std::any data);
> ```
would be used for

Reply:
we can do that too, thats fine

Reply:
then what should I name the functions?

Reply:
onMouseDown? I guess?